R/teal_lockfile.R

```diff
@@ -23,92 +44,139 @@
+    )
+  })
+  logger::log_trace("Lockfile creation started based on { getwd() }.")
+
+  process
 }
 
-create_renv_lockfile <- function(close = FALSE, lockfile_path = NULL) {
-  checkmate::assert_flag(close)
-  checkmate::assert_string(lockfile_path, na.ok = TRUE)
-  promise <- promises::future_promise({
-    # Below we can not use a file created in tempdir() directory.
-    # If a file is created in tempdir(), it gets deleted on 'then(onFulfilled' part.
-    shiny::onStop(function() file.remove(lockfile_path))
+#' @rdname teal_lockfile
+#' @keywords internal
+renv_snapshot <- function(lockfile_path = NULL, opts, sysenv, libpaths, wd) {
+  checkmate::assert_string(lockfile_path)
+  checkmate::assert_list(opts)
+  checkmate::assert_class(sysenv, "list")
+  checkmate::assert_character(libpaths, min.len = 1)
+  checkmate::assert_directory(wd)
 
-    renv_logs <- utils::capture.output(
-      renv::snapshot(
-        lockfile = lockfile_path,
-        prompt = FALSE,
-        force = TRUE
-        # type = is taken from renv::settings$snapshot.type()
-      )
+  # mirai starts in vanilla session in the R.home directory. We need to pass all session related info
+  options(opts)
+  do.call(Sys.setenv, sysenv)
+  .libPaths(libpaths)
+  setwd(wd)
+
+  out <- utils::capture.output(
+    renv <- renv::snapshot(
+      lockfile = lockfile_path,
+      prompt = FALSE,
+      force = TRUE
+      # type is taken from renv::settings$snapshot.type()
     )
-    if (any(grepl("Lockfile written", renv_logs))) {
-      logger::log_trace("lockfile created successfully.")
+  )
+
+  list(
+    out = out,
+    lockfile_path = lockfile_path,
+    length = length(renv$Packages)
+  )
+}
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_tracker <- function(process) {
+  checkmate::assert_class(process, "ExtendedTask", null.ok = TRUE)
+  if (inherits(process, "ExtendedTask")) {
+    tracker <- observeEvent(process$status(), {
+      if (process$status() != "running") {
+        teal_lockfile_handler(process)
+      }
+    })
+  } else {
+    # If it is not mirai proces, then lockfile was provided through options. Skip the process - file is ready.
+    shinyjs::show("teal-lockFile")
+  }
+}
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_handler <- function(process) {
+  renv_logs <- process$result()
+  if (any(grepl("Lockfile written to", renv_logs$out))) {
+    with <- if (any(grepl("WARNING:", renv_logs$out))) {
+      " with warning(s)"
+    } else if (any(grepl("ERROR:", renv_logs$out))) {
+      " with error(s)"
     } else {
-      logger::log_trace("lockfile created with issues.")
+      ""
     }
 
-    lockfile_path
-  })
-  if (close) {
-    # If the initial backend was only sequential, bring it back.
-    promises::then(promise, onFulfilled = function() {
-      future::plan(future::sequential)
-    })
+    logger::log_trace("Lockfile {renv_logs$lockfile_path} containing { renv_logs$length } packages created{ with }.")
```

Comment:
let's assume I got warnings or errors - how can I see them? 
I think we should log the contents of `renv_logs$out`.